### PR TITLE
Initial changes to move to pipelines

### PR DESCRIPTION
**Pull request by davidfowl**
- Change the Sockets abstraction from Channel\<byte[]\> to pipelines.

#615

### CODE REVIEW

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs [hunk 1]
```diff
@@ -199,6 +200,52 @@ public HttpConnectionDispatcherTests(ITestOutputHelper output) : base(output)
             }
         }
 
+        [InlineData(TransportType.LongPolling)]
+        [InlineData(TransportType.ServerSentEvents)]
+        [Theory]
+        public async Task PostSendsToConnection(TransportType transportType)
+        {
+            using (StartLog(out var loggerFactory, LogLevel.Debug))
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: Do something like `$"{nameof(PostSendsToConnection}_{transportType}"` for the test name

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs [hunk 2]
```diff
@@ -1229,7 +1276,14 @@ public class BlockingEndPoint : EndPoint
     {
         public override Task OnConnectedAsync(ConnectionContext connection)
         {
-            connection.Transport.Reader.WaitToReadAsync().Wait();
+            var wh = new ManualResetEventSlim();
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: What kind of variable name is wh?!

File: test/Microsoft.AspNetCore.Sockets.Tests/WebSocketsTests.cs [hunk 3]
```diff
@@ -54,10 +51,12 @@ public WebSocketsTests(ITestOutputHelper output)
...
 
-                    var buffer = await applicationSide.Reader.ReadAsync();
-                    Assert.Equal("Hello", Encoding.UTF8.GetString(buffer));
+                    var result = await connection.Transport.Input.ReadAsync();
+                    var buffer = result.Buffer;
+                    Assert.Equal("Hello", Encoding.UTF8.GetString(buffer.ToArray()));
+                    connection.Transport.Input.AdvanceTo(buffer.End);
 
-                    Assert.True(applicationSide.Writer.TryComplete());
+                    connection.Transport.Output.Commit();
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.Sockets.Tests/WebSocketsTests.cs]: What is `Commit` vs `Complete`?

File: src/Microsoft.AspNetCore.Sockets.Abstractions/Microsoft.AspNetCore.Sockets.Abstractions.csproj [hunk 4]
```diff
@@ -9,6 +9,9 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Http.Features" Version="$(MicrosoftAspNetCoreHttpFeaturesPackageVersion)" />
     <PackageReference Include="System.Threading.Channels" Version="$(SystemThreadingChannelsPackageVersion)" />
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Abstractions/Microsoft.AspNetCore.Sockets.Abstractions.csproj]: Is this reference still needed?

File: client-ts/FunctionalTests/EchoEndPoint.cs [hunk 5]
```diff
@@ -10,7 +11,20 @@ public class EchoEndPoint : EndPoint
...
+            var buffer = result.Buffer;
+
+            try
+            {
+                if (!buffer.IsEmpty)
+                {
+                    await connection.Transport.Output.WriteAsync(buffer.ToArray());
+                }
+            }
+            finally
```
Comment (muratg) [inline client-ts/FunctionalTests/EchoEndPoint.cs]: log exception?

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs [hunk 6]
```diff
@@ -19,6 +21,15 @@ public HubProtocolReaderWriter(IHubProtocol hubProtocol, IDataEncoder dataEncode
             _dataEncoder = dataEncoder;
         }
 
+        public bool ReadMessages(ReadOnlyBuffer<byte> buffer, IInvocationBinder binder, out IList<HubMessage> messages, out SequencePosition consumed, out SequencePosition examined)
+        {
+            // TODO: Fix this implementation to be incremental
```
Comment (muratg) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs]: Not clear to me why you wouldn't read to the end at once...

File: test/Microsoft.AspNetCore.Sockets.Tests/ConnectionManagerTests.cs [hunk 7]
```diff
@@ -188,15 +205,22 @@ public void ScanAfterDisposeNoops()
...
 
             appLifetime.Start();
 
             var connection = connectionManager.CreateConnection();
 
+            connection.Application.Output.OnReaderCompleted((error, state) =>
+            {
+                tcs.TrySetResult(null);
+
+            }, null);
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply test/Microsoft.AspNetCore.Sockets.Tests/ConnectionManagerTests.cs]: Not in a test

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs (hunk 2 above)
Comment (davidfowl) [reply test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: waitHandle

File: test/Microsoft.AspNetCore.Sockets.Tests/WebSocketsTests.cs (hunk 3 above)
Comment (davidfowl) [reply test/Microsoft.AspNetCore.Sockets.Tests/WebSocketsTests.cs]: Buggg

File: client-ts/FunctionalTests/EchoEndPoint.cs (hunk 5 above)
Comment (davidfowl) [reply client-ts/FunctionalTests/EchoEndPoint.cs]: No

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 8]
```diff
@@ -31,33 +30,36 @@ public LongPollingTransport(CancellationToken timeoutToken, ChannelReader<byte[]
         {
             try
             {
-                if (!await _application.WaitToReadAsync(token))
+                var result = await _application.ReadAsync(token);
+                var buffer = result.Buffer;
+
+                if (buffer.IsEmpty && result.IsCompleted)
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Then we write nothing.

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 9]
```diff
@@ -31,33 +30,36 @@ public LongPollingTransport(CancellationToken timeoutToken, ChannelReader<byte[]
...
-                context.Response.ContentLength = contentLength;
+                context.Response.ContentLength = buffer.Length;
                 context.Response.ContentType = "application/octet-stream";
 
-                foreach (var buffer in buffers)
+                try
+                {
+                    foreach (var segment in buffer)
+                    {
+                        MemoryMarshal.TryGetArray(segment, out var arraySegment);
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Will `Debug.Asset` We're using the managed memory pool so an array be used.

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs (hunk 1 above)
Comment (davidfowl) [reply test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: Other tests don't do this.

File: src/Microsoft.AspNetCore.Sockets.Abstractions/Microsoft.AspNetCore.Sockets.Abstractions.csproj (hunk 4 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/Microsoft.AspNetCore.Sockets.Abstractions.csproj]: Yes for the client APIs

File: samples/SocketsSample/EndPoints/MessagesEndPoint.cs [hunk 10]
```diff
@@ -20,15 +21,22 @@ public class MessagesEndPoint : EndPoint
 
             try
             {
-                while (await connection.Transport.Reader.WaitToReadAsync())
+                while (true)
                 {
-                    if (connection.Transport.Reader.TryRead(out var buffer))
+                    var result = await connection.Transport.Input.ReadAsync();
+                    var buffer = result.Buffer;
```
[the remark this answers is not in the document]
Comment (muratg) [reply samples/SocketsSample/EndPoints/MessagesEndPoint.cs]: Why nested `try`s?

**Pull request merged**